// @node-red/editor-api/lib/editor/index.js
/**
 * Copyright JS Foundation and other contributors, http://js.foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

var express = require("express");
var path = require('path');
const session = require('express-session');
var comms = require("./comms");
var library = require("./library");
var info = require("./settings");
const https = require('https')
var auth = require("../auth");
var nodes = require("../admin/nodes"); // TODO: move /icons into here
var needsPermission;
var runtimeAPI;
var log = require("@node-red/util").log;
var i18n = require("@node-red/util").i18n;
const fs = require('fs')
var apiUtil = require("../util");
var settingConfig = require('../../../../../Settings')


var ensureRuntimeStarted = function(req,res,next) {
    console.log(req.headers.authorization)
    req.session.flowid = "flowid1";
    req.session.token = req.headers.authorization;
    // req.session.token = req.headers.authorization;
    req.session.flowid = req.params.flowid;
    console.log("ensureRuntimeStarted: "+req.sessionID)
    runtimeAPI.isStarted().then( started => {
        if (!started) {
            log.error("Node-RED runtime not started");
            res.status(503).send("Not started");
        } else {
            if(req.session){
                console.log(settingConfig.getHostNodered()+':'+settingConfig.getPortNodered()+settingConfig.getUrlCheckToken())
                console.log(req.session.token)
                var token = req.session.token

                console.log(token)
                if (typeof token !== 'undefined'){

                    var http = require('http');
                    const options = {
                        hostname: settingConfig.getHostNodered(),
                        port: settingConfig.getPortNodered(),
                        path: settingConfig.getUrlCheckToken(),
                        method: 'GET',
                        headers: {
                            Authorization: `${token}`
                        }
                    }
                    callback = function(response) {
                        var str = ''
                        response.on('data', function (chunk) {
                            str += chunk;
                        });

                        response.on('end', function () {
                            // var rs = JSON.parse(str)
                            // activeFlow = rs;
                            // return rs
                            // console.log(str)
                            // console.log(rs.errorCode);
                            console.log(response.statusCode);
                            if (response.statusCode == 200) {
                                console.log("IN QUANG")
                                next()
                            }
                            else {
                                console.log("OUT QUANG")
                                res.status(405).send("Không thể xác thực!!!")
                            }
                        });
                    }
                    var req1 = http.request(options, callback);
                    req1.end();

                }
                else{
                    // fs.createReadStream('401unauthen.html').pipe(res);
                    res.status(401).send("Bạn chưa đăng nhập!!!")
                }

            }
            else {
                console.log("Session khong ton tai")
                res.status(401).send("Bạn chưa đăng nhập!!!")
            }
        }
    })
}

module.exports = {
    init: function(server, settings, _runtimeAPI) {
        runtimeAPI = _runtimeAPI;
        needsPermission = auth.needsPermission;
        if (!settings.disableEditor) {
            info.init(runtimeAPI);
            comms.init(server,settings,runtimeAPI);

            var ui = require("./ui");

            ui.init(runtimeAPI);

            var editorApp = express();
            if (settings.requireHttps === true) {

                editorApp.enable('trust proxy');
                editorApp.use(function (req, res, next) {
                    if (req.secure) {
                        next();
                    } else {
                        res.redirect('https://' + req.headers.host + req.originalUrl);
                    }
                });
            }
            if (settings.httpServerOptions) {
                for (var eOption in settings.httpServerOptions) {
                    editorApp.set(eOption, settings.httpServerOptions[eOption]);
                }
            }
            console.log("INNNNNNNNNNNNNN")
            editorApp.get("/",ensureRuntimeStarted,ui.ensureSlash,ui.editor);
            // editorApp.get("/",ensureRuntimeStarted,ui.ensureSlash,ui.editor);

            editorApp.get("/icons",needsPermission("nodes.read"),nodes.getIcons,apiUtil.errorHandler);
            editorApp.get("/icons/:module/:icon",ui.icon);
            editorApp.get("/icons/:scope/:module/:icon",ui.icon);

            editorApp.get(/^\/resources\/((?:@[^\/]+\/)?[^\/]+)\/(.+)$/,ui.moduleResource);

            var theme = require("./theme");
            theme.init(settings, runtimeAPI);
            editorApp.use("/theme",theme.app());


            editorApp.use("/",ui.editorResources);

            //Projects
            var projects = require("./projects");
            projects.init(runtimeAPI);
            editorApp.use("/projects",projects.app());0

            // Locales
            var locales = require("./locales");
            locales.init(runtimeAPI);
            editorApp.get(/^\/locales\/(.+)\/?$/,locales.get,apiUtil.errorHandler);

            // Library
            var library = require("./library");
            library.init(runtimeAPI);
            // editorApp.get("/library/:id",needsPermission("library.read"),library.getLibraryConfig);
            editorApp.get(/^\/library\/([^\/]+)\/([^\/]+)(?:$|\/(.*))/,needsPermission("library.read"),library.getEntry);
            editorApp.post(/^\/library\/([^\/]+)\/([^\/]+)\/(.*)/,needsPermission("library.write"),library.saveEntry);


            // Credentials
            var credentials = require("./credentials");
            credentials.init(runtimeAPI);
            editorApp.get('/credentials/:type/:id', needsPermission("credentials.read"),credentials.get,apiUtil.errorHandler);

            // Settings
            //  Main /settings route is an admin route - see lib/admin/settings.js
            // User Settings
            editorApp.get("/settings/user",needsPermission("settings.read"),info.userSettings,apiUtil.errorHandler);
            // User Settings
            editorApp.post("/settings/user",needsPermission("settings.write"),info.updateUserSettings,apiUtil.errorHandler);
            // SSH keys
            editorApp.use("/settings/user/keys",needsPermission("settings.write"),info.sshkeys());

            return editorApp;
        }
    },
    start: function() {
        var catalogPath = path.resolve(path.join(path.dirname(require.resolve("@node-red/editor-client")),"locales"));
        return i18n.registerMessageCatalogs([
            {namespace: "editor",   dir: catalogPath, file:"editor.json"},
            {namespace: "jsonata",  dir: catalogPath, file:"jsonata.json"},
            {namespace: "infotips", dir: catalogPath, file:"infotips.json"}
        ]).then(function(){
            comms.start();
        });
    },
    stop: comms.stop
}
